Callbacks separation, GtkProgressBar upgrade

skybon opened the pull request:
* Pilot attempt to move callbacks out of xqf.c. Refs #169
* Moved basic definitions to defs.h. It is now possible to include them without pulling xqf.h.
* GtkProgress calls ported to GtkProgressBar, existing GtkProgressBar calls made compliant with GTK2/3 specs.

illwieckz commented:
Hi @skybon, I like the `callbacks.c` idea!

For the moment, the code does not works, I get a segfault when I try to upate a master server list:

```c
#0  0x000000000046a31b in progress_bar_start (pbar=0xe42010 [GtkProgressBar], pulse_mode=1) at /home/illwieckz/dev/xqf-skybon/src/xqf-ui.c:653
#1  0x0000000000461f69 in stat_lists_state_handler (job=0xe5f1d0, state=STAT_UPDATE_SOURCE) at /home/illwieckz/dev/xqf-skybon/src/xqf.c:499
#2  0x000000000045a7b4 in change_state (job=0xe5f1d0, state=STAT_UPDATE_SOURCE) at /home/illwieckz/dev/xqf-skybon/src/stat.c:1755
#3  0x000000000045a9d9 in stat_next_masters (job=0xe5f1d0) at /home/illwieckz/dev/xqf-skybon/src/stat.c:1808
#4  0x000000000045ae2d in stat_next (job=0xe5f1d0) at /home/illwieckz/dev/xqf-skybon/src/stat.c:1913
#5  0x000000000045af51 in stat_start (job=0xe5f1d0) at /home/illwieckz/dev/xqf-skybon/src/stat.c:1939
#6  0x00000000004621bb in stat_lists (masters=0xd42970, names=0x0, servers=0x0, hosts=0x0) at /home/illwieckz/dev/xqf-skybon/src/xqf.c:559
#7  0x0000000000463460 in update_source_callback (widget=0xd2ac70 [GtkToolButton], data=0x0) at /home/illwieckz/dev/xqf-skybon/src/xqf.c:1013
#8  0x00007ffff6c65504 in _g_closure_invoke_va (closure=0xe42020, closure@entry=0xd515d0, return_value=0x50, return_value@entry=0x0, instance=0x1, instance@entry=0xd2ac70, args=0x0, args@entry=0x7fffffffd2d8, n_params=14975384, param_types=0x1) at /build/buildd/glib2.0-2.44.1/./gobject/gclosure.c:831
#9  0x00007ffff6c7efa7 in g_signal_emit_valist (instance=0xd2ac70, signal_id=<optimized out>, detail=0, var_args=0x7fffffffd2d8) at /build/buildd/glib2.0-2.44.1/./gobject/gsignal.c:3214
#10 0x00007ffff6c7fe4a in g_signal_emit_by_name (instance=0xe42020, detailed_signal=0x50 <error: Cannot access memory at address 0x50>) at /build/buildd/glib2.0-2.44.1/./gobject/gsignal.c:3401
#11 0x00007ffff6c65504 in _g_closure_invoke_va (closure=0xe42020, closure@entry=0xd5b1f0, return_value=0x50, return_value@entry=0x0, instance=0x1, instance@entry=0xd4c1d0, args=0x0, args@entry=0x7fffffffd5f0, n_params=14975384, param_types=0x1) at /build/buildd/glib2.0-2.44.1/./gobject/gclosure.c:831
#12 0x00007ffff6c7efa7 in g_signal_emit_valist (instance=0xd4c1d0, signal_id=<optimized out>, detail=0, var_args=var_args@entry=0x7fffffffd5f0) at /build/buildd/glib2.0-2.44.1/./gobject/gsignal.c:3214
#13 0x00007ffff6c7f8ff in g_signal_emit (instance=<optimized out>, signal_id=<optimized out>, detail=<optimized out>) at /build/buildd/glib2.0-2.44.1/./gobject/gsignal.c:3361
#14 0x00007ffff740c395 in gtk_real_button_released (button=0xd4c1d0 [GtkButton]) at /build/buildd/gtk+2.0-2.24.27/gtk/gtkbutton.c:1712
#18 0x00007ffff6c7f8ff in <emit signal ??? on instance 0xd4c1d0 [GtkButton]> (instance=<optimized out>, signal_id=<optimized out>, detail=<optimized out>) at /build/buildd/glib2.0-2.44.1/./gobject/gsignal.c:3361
#15 0x00007ffff6c652d5 in g_closure_invoke (closure=0x75cbb0, return_value=0x0, n_param_values=1, param_values=0x7fffffffd8a0, invocation_hint=0x7fffffffd840) at /build/buildd/glib2.0-2.44.1/./gobject/gclosure.c:768
#16 0x00007ffff6c76913 in signal_emit_unlocked_R (node=node@entry=0x7580c0, detail=detail@entry=0, instance=instance@entry=0xd4c1d0, emission_return=emission_return@entry=0x0, instance_and_params=instance_and_params@entry=0x7fffffffd8a0) at /build/buildd/glib2.0-2.44.1/./gobject/gsignal.c:3479
#17 0x00007ffff6c7f698 in g_signal_emit_valist (instance=<optimized out>, signal_id=<optimized out>, detail=<optimized out>, var_args=var_args@entry=0x7fffffffda20) at /build/buildd/glib2.0-2.44.1/./gobject/gsignal.c:3305
#19 0x00007ffff740b289 in gtk_button_button_release (widget=<optimized out>, event=<optimized out>) at /build/buildd/gtk+2.0-2.24.27/gtk/gtkbutton.c:1604
#24 0x00007ffff6c7f8ff in <emit signal ??? on instance 0xd4c1d0 [GtkButton]> (instance=instance@entry=0xd4c1d0, signal_id=<optimized out>, detail=detail@entry=0) at /build/buildd/glib2.0-2.44.1/./gobject/gsignal.c:3361
#20 0x00007ffff74b1d1f in _gtk_marshal_BOOLEAN__BOXED (closure=0x738050, return_value=0x7fffffffdc50, n_param_values=<optimized out>, param_values=0x7fffffffdd00, invocation_hint=<optimized out>, marshal_data=0x7ffff740b270 <gtk_button_button_release>) at gtkmarshalers.c:86
#21 0x00007ffff6c652d5 in g_closure_invoke (closure=0x738050, return_value=0x7fffffffdc50, n_param_values=2, param_values=0x7fffffffdd00, invocation_hint=0x7fffffffdca0) at /build/buildd/glib2.0-2.44.1/./gobject/gclosure.c:768
#22 0x00007ffff6c76f32 in signal_emit_unlocked_R (node=node@entry=0x7382c0, detail=detail@entry=0, instance=instance@entry=0xd4c1d0, emission_return=emission_return@entry=0x7fffffffddb0, instance_and_params=instance_and_params@entry=0x7fffffffdd00) at /build/buildd/glib2.0-2.44.1/./gobject/gsignal.c:3587
#23 0x00007ffff6c7f1a5 in g_signal_emit_valist (instance=<optimized out>, signal_id=<optimized out>, detail=<optimized out>, var_args=var_args@entry=0x7fffffffde90) at /build/buildd/glib2.0-2.44.1/./gobject/gsignal.c:3315
#25 0x00007ffff75c961c in gtk_widget_event_internal (widget=widget@entry=0xd4c1d0 [GtkButton], event=event@entry=0xe67df0) at /build/buildd/gtk+2.0-2.24.27/gtk/gtkwidget.c:5017
#26 0x00007ffff75c9911 in IA__gtk_widget_event (widget=widget@entry=0xd4c1d0 [GtkButton], event=event@entry=0xe67df0) at /build/buildd/gtk+2.0-2.24.27/gtk/gtkwidget.c:4814
#27 0x00007ffff74b0464 in IA__gtk_propagate_event (widget=0xd4c1d0 [GtkButton], event=0xe67df0) at /build/buildd/gtk+2.0-2.24.27/gtk/gtkmain.c:2520
#28 0x00007ffff74b08fb in IA__gtk_main_do_event (event=0xe67df0) at /build/buildd/gtk+2.0-2.24.27/gtk/gtkmain.c:1710
#29 0x00007ffff7123bac in gdk_event_dispatch (source=<optimized out>, callback=<optimized out>, user_data=<optimized out>) at /build/buildd/gtk+2.0-2.24.27/gdk/x11/gdkevents-x11.c:2425
#30 0x00007ffff698fc3d in g_main_context_dispatch (context=0x6f90e0) at /build/buildd/glib2.0-2.44.1/./glib/gmain.c:3122
#31 0x00007ffff698fc3d in g_main_context_dispatch (context=context@entry=0x6f90e0) at /build/buildd/glib2.0-2.44.1/./glib/gmain.c:3737
#32 0x00007ffff698ff20 in g_main_context_iterate (context=0x6f90e0, block=block@entry=1, dispatch=dispatch@entry=1, self=<optimized out>) at /build/buildd/glib2.0-2.44.1/./glib/gmain.c:3808
#33 0x00007ffff6990242 in g_main_loop_run (loop=0xe759f0) at /build/buildd/glib2.0-2.44.1/./glib/gmain.c:4002
#34 0x00007ffff74af857 in IA__gtk_main () at /build/buildd/gtk+2.0-2.24.27/gtk/gtkmain.c:1282
#35 0x0000000000468499 in main (argc=1, argv=0x7fffffffe308) at /home/illwieckz/dev/xqf-skybon/src/xqf.c:2756
```

skybon commented:
@illwieckz Memory error while manipulating a struct.

Old code used manual dumping of ProgressBar's GObject properties into struct. Interesting but error prone especially on GTK's internal changes.

Refactored and fixed.

illwieckz commented:
Well, Travis failed… But strangely, the first time I pulled your changes, it failed to build on my computer too with a similar error, then I just drop the branch and repulled it from scratch, and it builded nice…

These are the errors I get before I reclone your branch:

```gcc
In file included from /usr/include/glib-2.0/gobject/gobject.h:24:0,
                 from /usr/include/glib-2.0/gobject/gbinding.h:29,
                 from /usr/include/glib-2.0/glib-object.h:23,
                 from /usr/include/glib-2.0/gio/gioenums.h:28,
                 from /usr/include/glib-2.0/gio/giotypes.h:28,
                 from /usr/include/glib-2.0/gio/gio.h:26,
                 from /usr/include/gtk-2.0/gdk/gdkapplaunchcontext.h:30,
                 from /usr/include/gtk-2.0/gdk/gdk.h:32,
                 from /usr/include/gtk-2.0/gtk/gtk.h:32,
                 from /home/illwieckz/dev/xqf-skybon/src/xqf-ui.c:26:
/home/illwieckz/dev/xqf-skybon/src/xqf-ui.c: In function ‘progress_bar_start’:
/home/illwieckz/dev/xqf-skybon/src/xqf-ui.c:643:54: error: ‘user_data’ undeclared (first use in this function)
   gtk_progress_bar_set_pulse_step (GTK_PROGRESS_BAR (user_data), 0.03);
                                                      ^
/usr/include/glib-2.0/gobject/gtype.h:2186:57: note: in definition of macro ‘_G_TYPE_CIC’
     ((ct*) g_type_check_instance_cast ((GTypeInstance*) ip, gt))
                                                         ^
/usr/include/gtk-2.0/gtk/gtkprogressbar.h:41:43: note: in expansion of macro ‘G_TYPE_CHECK_INSTANCE_CAST’
 #define GTK_PROGRESS_BAR(obj)            (G_TYPE_CHECK_INSTANCE_CAST ((obj), GTK_TYPE_PROGRESS_BAR, GtkProgressBar))
                                           ^
/home/illwieckz/dev/xqf-skybon/src/xqf-ui.c:643:36: note: in expansion of macro ‘GTK_PROGRESS_BAR’
   gtk_progress_bar_set_pulse_step (GTK_PROGRESS_BAR (user_data), 0.03);
                                    ^
/home/illwieckz/dev/xqf-skybon/src/xqf-ui.c:643:54: note: each undeclared identifier is reported only once for each function it appears in
   gtk_progress_bar_set_pulse_step (GTK_PROGRESS_BAR (user_data), 0.03);
                                                      ^
/usr/include/glib-2.0/gobject/gtype.h:2186:57: note: in definition of macro ‘_G_TYPE_CIC’
     ((ct*) g_type_check_instance_cast ((GTypeInstance*) ip, gt))
                                                         ^
/usr/include/gtk-2.0/gtk/gtkprogressbar.h:41:43: note: in expansion of macro ‘G_TYPE_CHECK_INSTANCE_CAST’
 #define GTK_PROGRESS_BAR(obj)            (G_TYPE_CHECK_INSTANCE_CAST ((obj), GTK_TYPE_PROGRESS_BAR, GtkProgressBar))
                                           ^
/home/illwieckz/dev/xqf-skybon/src/xqf-ui.c:643:36: note: in expansion of macro ‘GTK_PROGRESS_BAR’
   gtk_progress_bar_set_pulse_step (GTK_PROGRESS_BAR (user_data), 0.03);
                                    ^
CMakeFiles/xqf.dir/build.make:767: recipe for target 'CMakeFiles/xqf.dir/src/xqf-ui.c.o' failed
make[2]: *** [CMakeFiles/xqf.dir/src/xqf-ui.c.o] Error 1
CMakeFiles/Makefile2:130: recipe for target 'CMakeFiles/xqf.dir/all' failed
make[1]: *** [CMakeFiles/xqf.dir/all] Error 2
Makefile:117: recipe for target 'all' failed
make: *** [all] Error 2
```

Travis did a similar thing:

```gcc
/home/travis/build/XQF/xqf/src/xqf-ui.c:643:54: error: use of undeclared
      identifier 'user_data'
                gtk_progress_bar_set_pulse_step (GTK_PROGRESS_BAR (user_...
                                                                   ^
/usr/include/gtk-2.0/gtk/gtkprogressbar.h:41:72: note: expanded from macro
      'GTK_PROGRESS_BAR'
#define GTK_PROGRESS_BAR(obj)            (G_TYPE_CHECK_INSTANCE_CAST ((obj), GTK...
                                                                       ^
/usr/include/glib-2.0/gobject/gtype.h:482:80: note: expanded from macro
      'G_TYPE_CHECK_INSTANCE_CAST'
  ...g_type, c_type)    (_G_TYPE_CIC ((instance), (g_type), c_type))
                                       ^
/usr/include/glib-2.0/gobject/gtype.h:1668:57: note: expanded from macro
      '_G_TYPE_CIC'
    ((ct*) g_type_check_instance_cast ((GTypeInstance*) ip, gt))
                                                        ^
1 error generated.
make[2]: *** [CMakeFiles/xqf.dir/src/xqf-ui.c.o] Error 1
make[1]: *** [CMakeFiles/xqf.dir/all] Error 2
make: *** [all] Error 2
```

I asked to Travis to retry…

illwieckz commented:
failed again…

skybon commented:
One of those 'oops' mistakes.
